lib/utils/gasEstimation.js

```diff
@@ -17,282 +15,178 @@ module.exports = (vm, runArgs, err, callback) => {
+    const base = index === 0;
+    let start = index;
+    let stop = 0;
+    let cost = 0;
+    let sixtyFloorths = 0;
+    const op = steps.ops[index];
+    const next = steps.ops[index + 1];
+    const intermediateCost = op.gasLeft - next.gasLeft;
+    let callingFee = fee || 0;
+    let compositeContext = false;
+
+    function addGas(val) {
+      if (sixtyFloorths) {
+        if (val >= sixtyFloorths) {
+          sixtyFloorths = 0;
+        } else {
+          sixtyFloorths -= val;
+        }
       }
-      return top + needed;
+      cost += val;
     }
-    return steps.ops[start].gasLeft - steps.ops[end].gasLeft;
-  };
-
-  /**
-   * Finds the index in systemOps index
-   * @param {Number} index index of op in steps.ops
-   */
-  const findRootScope = (opIndex) => {
-    const index = steps.findIndex(opIndex);
-    let begin = index;
-    let end = index;
-    const length = steps.systemOps.length;
-    while (begin && steps.systemOps[begin][1]) {
-      --begin;
-    }
-    while (end < length && steps.systemOps[end][1]) {
-      ++end;
-    }
-    return {
-      rootBegin: steps.systemOps[begin][0],
-      rootEnd: steps.systemOps[end - 1][0] + 1 // the op after the matching terminator
-    };
-  };
 
-  /**
-   * Find the range of the current system operation (CREATE, CALL, etc)
-   * @param {Number} index Index of the lowest gas point in the domain
-   */
-  const findParentScope = (opIndex) => {
-    let depth = steps.ops[opIndex].depth;
-    const index = steps.findIndex(opIndex);
-    let begin = index;
-    let end = index;
-    const length = steps.systemOps.length;
-    while (end < length && steps.systemOps[end][1] >= depth) {
-      ++end;
-    }
-    while (begin && steps.systemOps[begin][1] >= depth) {
-      --begin;
-    }
     return {
-      parentBegin: steps.systemOps[begin][0],
-      parentEnd: steps.systemOps[end - 1][0] + 1 // the op after the matching terminator
-    };
-  };
-
-  /**
-   * Finds the lowest index in a given search space
-   * @param {Number} start The beginning index of the search space
-   * @param {Number} stop The ending index of the search space
-   */
-  const findLowestIndex = (start, stop) => {
-    let lowestGas = Infinity;
-    let lowestIndex = stop;
-    for (let i = start; i <= stop; i++) {
-      if (steps.ops[i].gasLeft <= lowestGas) {
-        // Find the new lowest gas index
-        lowestIndex = i;
-        lowestGas = steps.ops[i].gasLeft;
+      start: () => start,
+      stop: () => stop,
+      setStart: (val) => {
+        start = val;
+        compositeContext = true;
+      },
+      setStop: (val) => {
+        stop = val;
+      },
+      getCost: () => ({ cost, sixtyFloorths }),
+      transfer: (ctx) => {
+        const values = ctx.getCost();
+        addGas(values.cost);
+        sixtyFloorths += values.sixtyFloorths;
+      },
+      addSixtyFloorth: (sixtyFloorth) => {
+        sixtyFloorths += sixtyFloorth;
+      },
+      addRange: (fee = false) => {
```

Comment:
Let's not (ab)use javascript's untyped types here and set proper defaults for the `fee` case. Probably just a matter of `if(fee !== undefined) {`. Though I _suspect_ that changing the method signature to something like `(fee = 0, useFee = false)` would be nicer to V8's optimizer.

Actually, just creating specialized methods for the ~2ish cases here would work well? /shrug

In any case, some light comments clarifying what each branch is for would be nice. :-)